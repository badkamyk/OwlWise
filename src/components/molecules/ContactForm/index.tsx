import emailjs from "@emailjs/browser";
import { zodResolver } from "@hookform/resolvers/zod";
import { useRef } from "react";
import { FormProvider, SubmitHandler, useForm } from "react-hook-form";
import { z } from "zod";
import HeadingButton from "../../atoms/HeadingButton";
import TextArea from "../../atoms/TextArea";
import CompleteInput from "../CompleteInput";

const formSchema = z.object({
	email: z.string().email("Email must be valid"),
	subject: z.string().min(3, "Subject must be at least 3 characters"),
	message: z.string().min(10, "Message must be at least 10 characters"),
});

type FormSchemaType = z.infer<typeof formSchema>;

export default function ContactForm() {
	const methods = useForm<FormSchemaType>({
		resolver: zodResolver(formSchema),
	});
	const formRef = useRef<HTMLFormElement>(null);
	const onSubmit: SubmitHandler<FormSchemaType> = () => {
		emailjs
			.sendForm(
				"YOUR_SERVICE_ID",
				"YOUR_TEMPLATE_ID",
				formRef.current as string | HTMLFormElement,
				"YOUR_PUBLIC_KEY"
			)
			.then(
				result => {
					console.log(result.text);
				},
				error => {
					console.log(error.text);
				}
			);
		methods.reset();
	};

	return (
		<FormProvider {...methods}>
			<form
				ref={formRef}
				className="space-y-8"
				onSubmit={methods.handleSubmit(onSubmit)}
			>
				<CompleteInput
					type={"email"}
					id={"email"}
					placeholder={"youremail@example.com"}
					text={"Your email"}
					required={true}
				/>
				<span className="mt-2 block text-red-600">
					{methods.formState.errors.email?.message}
				</span>
				<CompleteInput
					type={"text"}
					id={"subject"}
					placeholder={"Your subject"}
					text={"Subject"}
					required={true}
				/>
				<span className="mt-2 block text-red-600">
					{methods.formState.errors.subject?.message}
				</span>
				<TextArea
					id={"message"}
					placeholder={"Your message..."}
					required={true}
				/>
				<span className="mt-2 block text-red-600">
					{methods.formState.errors.message?.message}
				</span>
				<HeadingButton
					text={"Send message"}
					type={"submit"}
					disabled={methods.formState.isSubmitting}
				/>
			</form>
		</FormProvider>
	);
}
