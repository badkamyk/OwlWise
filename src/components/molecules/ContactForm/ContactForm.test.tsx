import { fireEvent, waitFor } from "@testing-library/dom";
import { act, render, renderHook, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import { describe, expect, it, vi } from "vitest";
import ContactForm from "./index";

describe("ContactForm", () => {
	const onSubmit = vi.fn();
	it("renders ContactForm correctly", () => {
		const { result } = renderHook(() => useForm());
		const { container } = render(
			<FormProvider {...result.current}>
				<ContactForm onSubmit={onSubmit} />
			</FormProvider>
		);
		expect(screen.getAllByRole("textbox")).toHaveLength(3);
		expect(screen.getByPlaceholderText("youremail@example.com")).toBeTruthy();
		expect(screen.getByText("Your email")).toBeTruthy();
		expect(screen.getByPlaceholderText("Your subject")).toBeTruthy();
		expect(screen.getByText("Subject")).toBeTruthy();
		expect(screen.getByPlaceholderText("Your message...")).toBeTruthy();
		expect(screen.getByText("Your message")).toBeTruthy();
		expect(screen.getByRole("button")).toHaveProperty("type", "submit");
		expect(container).toMatchSnapshot();
	});

	it("Calls onSubmit after submitting the form", async () => {
		const { result } = renderHook(() => useForm());
		render(
			<FormProvider {...result.current}>
				<ContactForm onSubmit={onSubmit} />
			</FormProvider>
		);
		fireEvent.input(screen.getByPlaceholderText("youremail@example.com"), {
			target: {
				value: "email@example.com",
			},
		});
		fireEvent.input(screen.getByPlaceholderText("Your subject"), {
			target: {
				value: "subject",
			},
		});
		fireEvent.input(screen.getByPlaceholderText("Your message..."), {
			target: {
				value: "this is a test message from user",
			},
		});
		await act(() => {
			fireEvent.submit(screen.getByTestId("contact-form"));
		});

		await waitFor(() => expect(onSubmit).toHaveBeenCalled());
	});
});
