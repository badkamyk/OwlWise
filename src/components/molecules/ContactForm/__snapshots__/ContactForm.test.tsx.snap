// Vitest Snapshot v1

exports[`ContactForm > renders ContactForm correctly 1`] = `
<div>
  <form
    class="space-y-8"
    data-testid="contact-form"
  >
    <div>
      <label
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-gray-300"
        for="email"
      >
        Your email
      </label>
      <input
        aria-invalid="false"
        class="focus:ring-primary-500 focus:border-primary-500 dark:focus:ring-primary-500 dark:focus:border-primary-500 dark:shadow-sm-light block w-full rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 shadow-sm dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400"
        id="email"
        name="email"
        placeholder="youremail@example.com"
        required=""
        type="email"
      />
    </div>
    <div>
      <label
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-gray-300"
        for="subject"
      >
        Subject
      </label>
      <input
        aria-invalid="false"
        class="focus:ring-primary-500 focus:border-primary-500 dark:focus:ring-primary-500 dark:focus:border-primary-500 dark:shadow-sm-light block w-full rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 shadow-sm dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400"
        id="subject"
        name="subject"
        placeholder="Your subject"
        required=""
        type="text"
      />
    </div>
    <div
      class="sm:col-span-2"
    >
      <label
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-gray-300"
        for="message"
      >
        Your message
      </label>
      <textarea
        aria-invalid="false"
        class="focus:ring-primary-500 focus:border-primary-500 dark:focus:ring-primary-500 dark:focus:border-primary-500 block w-full rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 shadow-sm dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400"
        id="message"
        name="message"
        placeholder="Your message..."
        required=""
      />
    </div>
    <button
      class="group relative mb-2 mr-2 inline-flex items-center justify-center overflow-hidden rounded-lg bg-gradient-to-br from-green-400 to-blue-600 p-0.5 text-sm font-medium text-gray-900 hover:text-white focus:outline-none focus:ring-4 focus:ring-green-200 group-hover:from-green-400 group-hover:to-blue-600 dark:text-white dark:focus:ring-green-800"
      data-testid="heading-button"
      type="submit"
    >
      <a
        class="relative rounded-md bg-white px-5 py-2.5 transition-all duration-75 ease-in group-hover:bg-opacity-0 dark:bg-gray-900"
      >
        Send message
      </a>
    </button>
  </form>
  )
</div>
`;
